crates/planner/src/logical/ty.rs

```diff
@@ -0,0 +1,243 @@
+/// Hence we must fully resolve each [TyId] when testing for equality.
+impl<'a> PartialEq for TypeWithCtx<'a> {
+    fn eq(&self, other: &Self) -> bool {
+        match (self.0, other.0) {
+            (Type::Var(a), Type::Var(b)) | (Type::Row(a), Type::Row(b)) => {
+                a.len() == b.len()
+                    && a.iter()
+                        .enumerate()
+                        .all(|(i, (name, id))| name == &b[i].0 && id.with_ctx(self.1) == b[i].1.with_ctx(other.1))
+            }
+            (Type::Tup(a), Type::Tup(b)) => {
+                a.len() == b.len()
+                    && a.iter()
+                        .enumerate()
+                        .all(|(i, id)| id.with_ctx(self.1) == b[i].with_ctx(other.1))
+            }
+            (Type::Alg(a), Type::Alg(b)) => a == b,
+            _ => false,
+        }
+    }
+}
+
+impl<'a> Display for TypeWithCtx<'a> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let Self(ty, ctx) = self;
+        match ty {
+            Type::Alg(ty) => write!(f, "{}", fmt_algebraic_type(ty)),
+            Type::Var(fields) | Type::Row(fields) => {
+                write!(f, "(")?;
+                let (name, id) = &fields[0];
+                write!(f, "{}: {}", name, id.with_ctx(ctx))?;
+                for (name, id) in &fields[1..] {
+                    write!(f, ", {}: {}", name, id.with_ctx(ctx))?;
+                }
+                write!(f, ")")
+            }
+            Type::Tup(types) => {
+                write!(f, "(")?;
+                write!(f, "{}", types[0].with_ctx(ctx))?;
+                for id in &types[1..] {
+                    write!(f, ", {}", id.with_ctx(ctx))?;
+                }
+                write!(f, ")")
+            }
+        }
+    }
+}
+
+/// The type of a relation or scalar expression
+pub enum Type {
+    /// A base relation
+    Var(Box<[(String, TyId)]>),
+    /// A derived relation
+    Row(Box<[(String, TyId)]>),
+    /// A join relation
+    Tup(Box<[TyId]>),
+    /// A column type
+    Alg(AlgebraicType),
+}
+
+static_assert_size!(Type, 24);
+
+impl Type {
+    /// A constant for the bool type
+    pub const BOOL: Self = Self::Alg(AlgebraicType::Bool);
+
+    /// A constant for the string type
+    pub const STR: Self = Self::Alg(AlgebraicType::String);
+
+    /// Wrap this type with its typing context
+    pub fn with_ctx<'a>(&'a self, ctx: &'a TyCtx) -> TypeWithCtx {
+        TypeWithCtx(self, ctx)
+    }
+
+    /// Is this type compatible with this binary operator?
+    pub fn is_compatible_with(&self, op: BinOp) -> bool {
+        match (op, self) {
+            // Logic operators take booleans
+            (BinOp::And | BinOp::Or, Type::Alg(AlgebraicType::Bool)) => true,
+            // Comparison operators take integers or floats
+            (BinOp::Lt | BinOp::Gt | BinOp::Lte | BinOp::Gte, Type::Alg(t)) => t.is_integer() || t.is_float(),
+            // Equality supports numerics, strings, and bytes
```

Comment:
Why can't the others?

Reply:
Mainly to avoid inconsistencies with other type systems. In the case of Rust, sums and products correspond to enum and struct types. Those types could have custom Eq and Ord implementations and as of right now, we don't have a way of representing that in SpacetimeDB.